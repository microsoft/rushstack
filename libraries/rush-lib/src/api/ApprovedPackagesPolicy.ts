// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as path from 'node:path';

import { ApprovedPackagesConfiguration } from './ApprovedPackagesConfiguration';
import { RushConstants } from '../logic/RushConstants';
import type {
  RushConfiguration,
  IRushConfigurationJson,
  IApprovedPackagesPolicyJson
} from './RushConfiguration';

/**
 * This is a helper object for RushConfiguration.
 * It exposes the "approvedPackagesPolicy" feature from rush.json.
 * @public
 */
export class ApprovedPackagesPolicy {
  /**
   * Whether the feature is enabled.  The feature is enabled if the "approvedPackagesPolicy"
   * field is assigned in rush.json.
   */
  public readonly enabled: boolean;

  /**
   * A list of NPM package scopes that will be excluded from review (e.g. `@types`)
   */
  public readonly ignoredNpmScopes: ReadonlySet<string>;

  /**
   * A list of category names that are valid for usage as the RushConfigurationProject.reviewCategory field.
   * This array will never be undefined.
   */
  public readonly reviewCategories: ReadonlySet<string>;

  /**
   * Packages approved for usage in a web browser.  This is the stricter of the two types, so by default
   * all new packages are added to this file.
   *
   * @remarks
   *
   * This is part of an optional approval workflow, whose purpose is to review any new dependencies
   * that are introduced (e.g. maybe a legal review is required, or maybe we are trying to minimize bloat).
   * When Rush discovers a new dependency has been added to package.json, it will update the file.
   * The intent is that the file will be stored in Git and tracked by a branch policy that notifies
   * reviewers when a PR attempts to modify the file.
   *
   * Example filename: `C:\MyRepo\common\config\rush\browser-approved-packages.json`
   */
  public readonly browserApprovedPackages: ApprovedPackagesConfiguration;

  /**
   * Packages approved for usage everywhere *except* in a web browser.
   *
   * @remarks
   *
   * This is part of an optional approval workflow, whose purpose is to review any new dependencies
   * that are introduced (e.g. maybe a legal review is required, or maybe we are trying to minimize bloat).
   * The intent is that the file will be stored in Git and tracked by a branch policy that notifies
   * reviewers when a PR attempts to modify the file.
   *
   * Example filename: `C:\MyRepo\common\config\rush\browser-approved-packages.json`
   */
  public readonly nonbrowserApprovedPackages: ApprovedPackagesConfiguration;

  /** @internal */
  public constructor(rushConfiguration: RushConfiguration, rushConfigurationJson: IRushConfigurationJson) {
    const approvedPackagesPolicy: IApprovedPackagesPolicyJson =
      rushConfigurationJson.approvedPackagesPolicy || {};

    this.enabled = !!rushConfigurationJson.approvedPackagesPolicy;
    this.ignoredNpmScopes = new Set<string>(approvedPackagesPolicy.ignoredNpmScopes);
    this.reviewCategories = new Set<string>(approvedPackagesPolicy.reviewCategories);

    if (this.enabled) {
      if (!this.reviewCategories.size) {
        throw new Error(
          `The "approvedPackagesPolicy" feature is enabled ${RushConstants.rushJsonFilename}, but the reviewCategories` +
            ` list is not configured.`
        );
      }
    }

    // Load browser-approved-packages.json
    const browserApprovedPackagesPath: string = path.join(
      rushConfiguration.commonRushConfigFolder,
      RushConstants.browserApprovedPackagesFilename
    );
    this.browserApprovedPackages = new ApprovedPackagesConfiguration(browserApprovedPackagesPath);
    this.browserApprovedPackages.tryLoadFromFile(this.enabled);

    // Load nonbrowser-approved-packages.json
    const nonbrowserApprovedPackagesPath: string = path.join(
      rushConfiguration.commonRushConfigFolder,
      RushConstants.nonbrowserApprovedPackagesFilename
    );
    this.nonbrowserApprovedPackages = new ApprovedPackagesConfiguration(nonbrowserApprovedPackagesPath);
    this.nonbrowserApprovedPackages.tryLoadFromFile(this.enabled);
  }
}
