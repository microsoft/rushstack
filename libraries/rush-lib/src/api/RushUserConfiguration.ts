// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { FileSystem, JsonFile, JsonSchema } from '@rushstack/node-core-library';
import * as path from 'path';

import { Utilities } from '../utilities/Utilities';
import { RushConstants } from '../logic/RushConstants';
import schemaJson from '../schemas/rush-user-settings.schema.json';

interface IRushUserSettingsJson {
  buildCacheFolder?: string;
}

/**
 * Rush per-user configuration data.
 *
 * @beta
 */
export class RushUserConfiguration {
  private static _schema: JsonSchema = JsonSchema.fromLoadedObject(schemaJson);

  /**
   * If provided, store build cache in the specified folder. Must be an absolute path.
   */
  public readonly buildCacheFolder: string | undefined;

  private constructor(rushUserConfigurationJson: IRushUserSettingsJson | undefined) {
    this.buildCacheFolder = rushUserConfigurationJson?.buildCacheFolder;
    if (this.buildCacheFolder && !path.isAbsolute(this.buildCacheFolder)) {
      throw new Error('buildCacheFolder must be an absolute path');
    }
  }

  public static async initializeAsync(): Promise<RushUserConfiguration> {
    const rushUserFolderPath: string = RushUserConfiguration.getRushUserFolderPath();
    const rushUserSettingsFilePath: string = path.join(rushUserFolderPath, 'settings.json');
    let rushUserSettingsJson: IRushUserSettingsJson | undefined;
    try {
      rushUserSettingsJson = await JsonFile.loadAndValidateAsync(
        rushUserSettingsFilePath,
        RushUserConfiguration._schema
      );
    } catch (e) {
      if (!FileSystem.isNotExistError(e as Error)) {
        throw e;
      }
    }

    return new RushUserConfiguration(rushUserSettingsJson);
  }

  public static getRushUserFolderPath(): string {
    const homeFolderPath: string = Utilities.getHomeFolder();
    return `${homeFolderPath}/${RushConstants.rushUserConfigurationFolderName}`;
  }
}
