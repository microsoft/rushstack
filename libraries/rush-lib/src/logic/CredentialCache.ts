// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { FileSystem, JsonFile, JsonSchema, LockFile } from '@rushstack/node-core-library';

import { Utilities } from '../utilities/Utilities';
import { RushUserConfiguration } from '../api/RushUserConfiguration';
import schemaJson from '../schemas/credentials.schema.json';
import { objectsAreDeepEqual } from '../utilities/objectUtilities';

const CACHE_FILENAME: string = 'credentials.json';
const LATEST_CREDENTIALS_JSON_VERSION: string = '0.1.0';

interface ICredentialCacheJson {
  version: string;
  cacheEntries: {
    [credentialCacheId: string]: ICacheEntryJson;
  };
}

interface ICacheEntryJson {
  expires: number;
  credential: string;
  credentialMetadata?: object;
}

/**
 * @beta
 */
export interface ICredentialCacheEntry {
  expires?: Date;
  credential: string;
  credentialMetadata?: object;
}

/**
 * @beta
 */
export interface ICredentialCacheOptions {
  supportEditing: boolean;
}

/**
 * @beta
 */
export class CredentialCache /* implements IDisposable */ {
  private readonly _cacheFilePath: string;
  private readonly _cacheEntries: Map<string, ICacheEntryJson>;
  private _modified: boolean = false;
  private _disposed: boolean = false;
  private _supportsEditing: boolean;
  private readonly _lockfile: LockFile | undefined;

  private constructor(
    cacheFilePath: string,
    loadedJson: ICredentialCacheJson | undefined,
    lockfile: LockFile | undefined
  ) {
    if (loadedJson && loadedJson.version !== LATEST_CREDENTIALS_JSON_VERSION) {
      throw new Error(`Unexpected credentials.json file version: ${loadedJson.version}`);
    }

    this._cacheFilePath = cacheFilePath;
    this._cacheEntries = new Map<string, ICacheEntryJson>(Object.entries(loadedJson?.cacheEntries || {}));
    this._supportsEditing = !!lockfile;
    this._lockfile = lockfile;
  }

  public static async initializeAsync(options: ICredentialCacheOptions): Promise<CredentialCache> {
    const rushUserFolderPath: string = RushUserConfiguration.getRushUserFolderPath();
    const cacheFilePath: string = `${rushUserFolderPath}/${CACHE_FILENAME}`;
    const jsonSchema: JsonSchema = JsonSchema.fromLoadedObject(schemaJson);

    let loadedJson: ICredentialCacheJson | undefined;
    try {
      loadedJson = await JsonFile.loadAndValidateAsync(cacheFilePath, jsonSchema);
    } catch (e) {
      if (!FileSystem.isErrnoException(e as Error)) {
        throw e;
      }
    }

    let lockfile: LockFile | undefined;
    if (options.supportEditing) {
      lockfile = await LockFile.acquireAsync(rushUserFolderPath, `${CACHE_FILENAME}.lock`);
    }

    const credentialCache: CredentialCache = new CredentialCache(cacheFilePath, loadedJson, lockfile);
    return credentialCache;
  }

  public static async usingAsync(
    options: ICredentialCacheOptions,
    doActionAsync: (credentialCache: CredentialCache) => Promise<void> | void
  ): Promise<void> {
    await Utilities.usingAsync(async () => await CredentialCache.initializeAsync(options), doActionAsync);
  }

  public setCacheEntry(cacheId: string, entry: ICredentialCacheEntry): void {
    this._validate(true);

    const { expires, credential, credentialMetadata } = entry;
    const expiresMilliseconds: number = expires?.getTime() || 0;
    const existingCacheEntry: ICacheEntryJson | undefined = this._cacheEntries.get(cacheId);
    if (
      existingCacheEntry?.credential !== credential ||
      existingCacheEntry?.expires !== expiresMilliseconds ||
      !objectsAreDeepEqual(existingCacheEntry?.credentialMetadata, credentialMetadata)
    ) {
      this._modified = true;
      this._cacheEntries.set(cacheId, {
        expires: expiresMilliseconds,
        credential,
        credentialMetadata
      });
    }
  }

  public tryGetCacheEntry(cacheId: string): ICredentialCacheEntry | undefined {
    this._validate(false);

    const cacheEntry: ICacheEntryJson | undefined = this._cacheEntries.get(cacheId);
    if (cacheEntry) {
      const result: ICredentialCacheEntry = {
        expires: cacheEntry.expires ? new Date(cacheEntry.expires) : undefined,
        credential: cacheEntry.credential,
        credentialMetadata: cacheEntry.credentialMetadata
      };

      return result;
    } else {
      return undefined;
    }
  }

  public deleteCacheEntry(cacheId: string): void {
    this._validate(true);

    if (this._cacheEntries.has(cacheId)) {
      this._modified = true;
      this._cacheEntries.delete(cacheId);
    }
  }

  public trimExpiredEntries(): void {
    this._validate(true);

    const now: number = Date.now();
    for (const [cacheId, cacheEntry] of this._cacheEntries.entries()) {
      if (cacheEntry.expires < now) {
        this._cacheEntries.delete(cacheId);
        this._modified = true;
      }
    }
  }

  public async saveIfModifiedAsync(): Promise<void> {
    this._validate(true);

    if (this._modified) {
      const cacheEntriesJson: { [cacheId: string]: ICacheEntryJson } = {};
      for (const [cacheId, cacheEntry] of this._cacheEntries.entries()) {
        cacheEntriesJson[cacheId] = cacheEntry;
      }

      const newJson: ICredentialCacheJson = {
        version: LATEST_CREDENTIALS_JSON_VERSION,
        cacheEntries: cacheEntriesJson
      };
      await JsonFile.saveAsync(newJson, this._cacheFilePath, {
        ensureFolderExists: true,
        updateExistingFile: true,
        ignoreUndefinedValues: true
      });

      this._modified = false;
    }
  }

  public dispose(): void {
    this._lockfile?.release();
    this._disposed = true;
  }

  private _validate(requiresEditing: boolean): void {
    if (!this._supportsEditing && requiresEditing) {
      throw new Error(`This instance of ${CredentialCache.name} does not support editing.`);
    }

    if (this._disposed) {
      throw new Error(`This instance of ${CredentialCache.name} has been disposed.`);
    }
  }
}
