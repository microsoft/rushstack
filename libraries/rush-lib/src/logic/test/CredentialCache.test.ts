// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { LockFile, Async, FileSystem } from '@rushstack/node-core-library';
import { RushUserConfiguration } from '../../api/RushUserConfiguration';
import { CredentialCache, type ICredentialCacheOptions } from '../CredentialCache';

const FAKE_RUSH_USER_FOLDER: string = '~/.rush-user';

interface IPathsTestCase extends Required<Pick<ICredentialCacheOptions, 'cacheFilePath'>> {
  testCaseName: string;
}

describe(CredentialCache.name, () => {
  let fakeFilesystem: { [key: string]: string };
  let filesystemLocks: { [key: string]: Promise<void> };
  let unresolvedLockfiles: Set<LockFile>;

  beforeEach(() => {
    fakeFilesystem = {};
    filesystemLocks = {};
    unresolvedLockfiles = new Set();
  });

  beforeEach(() => {
    jest.spyOn(RushUserConfiguration, 'getRushUserFolderPath').mockReturnValue(FAKE_RUSH_USER_FOLDER);

    // TODO: Consider expanding these mocks and moving them to node-core-library
    jest
      .spyOn(LockFile, 'acquire')
      .mockImplementation(async (folderPath: string, lockFilePath: string, maxWaitMs?: number) => {
        const fullPath: string = `${folderPath}/${lockFilePath}`;
        const existingLock: Promise<void> | undefined = filesystemLocks[fullPath];
        if (existingLock) {
          if (maxWaitMs === undefined) {
            await existingLock;
          } else {
            await Promise.race([existingLock, Async.sleepAsync(maxWaitMs)]);
          }
        }

        let release: () => void;
        const lockPromise: Promise<void> = new Promise<void>((resolve: () => void) => {
          release = resolve;
        });

        // eslint-disable-next-line require-atomic-updates
        filesystemLocks[fullPath] = lockPromise;
        const result: LockFile = {
          release: () => {
            release();
            unresolvedLockfiles.delete(result);
          }
        } as LockFile;
        unresolvedLockfiles.add(result);
        return result;
      });

    jest
      .spyOn(FileSystem, 'writeFileAsync')
      .mockImplementation(async (filePath: string, data: Buffer | string) => {
        fakeFilesystem[filePath] = data.toString();
      });

    jest.spyOn(FileSystem, 'readFileAsync').mockImplementation(async (filePath: string) => {
      if (filePath in fakeFilesystem) {
        return fakeFilesystem[filePath];
      } else {
        const notExistError: NodeJS.ErrnoException = new Error(
          `ENOENT: no such file or directory, open '${filePath}'`
        );
        notExistError.code = 'ENOENT';
        notExistError.errno = -2;
        notExistError.syscall = 'open';
        notExistError.path = filePath;
        throw notExistError;
      }
    });
  });

  afterEach(() => {
    for (const lockfile of unresolvedLockfiles) {
      lockfile.release();
    }
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  describe.each<IPathsTestCase>([
    {
      testCaseName: 'default cache path',
      cacheFilePath: `${FAKE_RUSH_USER_FOLDER}/credentials.json`
    },
    {
      testCaseName: 'custom cache path with no suffix',
      cacheFilePath: `${FAKE_RUSH_USER_FOLDER}/my-cache-name`
    },
    {
      testCaseName: 'custom cache path with json suffix',
      cacheFilePath: `${FAKE_RUSH_USER_FOLDER}/my-cache-name.json`
    }
  ])('cache paths [$testCaseName]', ({ cacheFilePath }) => {
    it("initializes a credential cache correctly when one doesn't exist on disk", async () => {
      const credentialCache: CredentialCache = await CredentialCache.initializeAsync({
        supportEditing: false
      });
      expect(credentialCache).toBeDefined();
      credentialCache.dispose();
    });

    it('initializes a credential cache correctly when one exists on disk', async () => {
      const credentialId: string = 'test-credential';
      const credentialValue: string = 'test-value';
      fakeFilesystem[cacheFilePath] = JSON.stringify({
        version: '0.1.0',
        cacheEntries: {
          [credentialId]: {
            expires: 0,
            credential: credentialValue
          }
        }
      });

      const credentialCache: CredentialCache = await CredentialCache.initializeAsync({
        cacheFilePath: cacheFilePath,
        supportEditing: false
      });
      expect(credentialCache.tryGetCacheEntry(credentialId)?.credential).toEqual(credentialValue);
      expect(credentialCache.tryGetCacheEntry(credentialId)?.expires).toBeUndefined();
      credentialCache.dispose();
    });

    it('initializes a credential cache correctly when one exists on disk with a expired credential', async () => {
      const credentialId: string = 'test-credential';
      const credentialValue: string = 'test-value';
      fakeFilesystem[cacheFilePath] = JSON.stringify({
        version: '0.1.0',
        cacheEntries: {
          [credentialId]: {
            expires: 100, // Expired
            credential: credentialValue
          }
        }
      });

      const credentialCache: CredentialCache = await CredentialCache.initializeAsync({
        cacheFilePath: cacheFilePath,
        supportEditing: false
      });
      expect(credentialCache.tryGetCacheEntry(credentialId)?.credential).toEqual(credentialValue);
      expect(credentialCache.tryGetCacheEntry(credentialId)?.expires).toMatchSnapshot('expiration');
      credentialCache.dispose();
    });

    it('correctly trims expired credentials', async () => {
      const credentialId: string = 'test-credential';
      const credentialValue: string = 'test-value';
      fakeFilesystem[cacheFilePath] = JSON.stringify({
        version: '0.1.0',
        cacheEntries: {
          [credentialId]: {
            expires: 100, // Expired
            credential: credentialValue
          }
        }
      });

      const credentialCache: CredentialCache = await CredentialCache.initializeAsync({
        cacheFilePath: cacheFilePath,
        supportEditing: true
      });
      credentialCache.trimExpiredEntries();
      expect(credentialCache.tryGetCacheEntry(credentialId)).toBeUndefined();
      await credentialCache.saveIfModifiedAsync();
      credentialCache.dispose();

      expect(fakeFilesystem[cacheFilePath]).toMatchSnapshot('credential cache file');
    });

    it('correctly adds a new credential', async () => {
      const credentialId: string = 'test-credential';
      const credentialValue: string = 'test-value';

      const credentialCache1: CredentialCache = await CredentialCache.initializeAsync({
        cacheFilePath: cacheFilePath,
        supportEditing: true
      });
      credentialCache1.setCacheEntry(credentialId, { credential: credentialValue });
      expect(credentialCache1.tryGetCacheEntry(credentialId)?.credential).toEqual(credentialValue);
      expect(credentialCache1.tryGetCacheEntry(credentialId)?.expires).toBeUndefined();
      await credentialCache1.saveIfModifiedAsync();
      credentialCache1.dispose();

      expect(fakeFilesystem[cacheFilePath]).toMatchSnapshot('credential cache file');

      const credentialCache2: CredentialCache = await CredentialCache.initializeAsync({
        cacheFilePath: cacheFilePath,
        supportEditing: false
      });
      expect(credentialCache2.tryGetCacheEntry(credentialId)?.credential).toEqual(credentialValue);
      expect(credentialCache2.tryGetCacheEntry(credentialId)?.expires).toBeUndefined();
      credentialCache2.dispose();
    });

    it('correctly updates an existing credential', async () => {
      const credentialId: string = 'test-credential';
      const credentialValue: string = 'test-value';
      const newCredentialValue: string = 'new-test-value';
      fakeFilesystem[cacheFilePath] = JSON.stringify({
        version: '0.1.0',
        cacheEntries: {
          [credentialId]: {
            expires: 0,
            credential: credentialValue
          }
        }
      });

      const credentialCache1: CredentialCache = await CredentialCache.initializeAsync({
        cacheFilePath: cacheFilePath,
        supportEditing: true
      });
      credentialCache1.setCacheEntry(credentialId, { credential: newCredentialValue });
      expect(credentialCache1.tryGetCacheEntry(credentialId)?.credential).toEqual(newCredentialValue);
      expect(credentialCache1.tryGetCacheEntry(credentialId)?.expires).toBeUndefined();
      await credentialCache1.saveIfModifiedAsync();
      credentialCache1.dispose();

      expect(fakeFilesystem[cacheFilePath]).toMatchSnapshot('credential cache file');

      const credentialCache2: CredentialCache = await CredentialCache.initializeAsync({
        cacheFilePath: cacheFilePath,
        supportEditing: false
      });
      expect(credentialCache2.tryGetCacheEntry(credentialId)?.credential).toEqual(newCredentialValue);
      expect(credentialCache2.tryGetCacheEntry(credentialId)?.expires).toBeUndefined();
      credentialCache2.dispose();
    });

    it('correctly deletes an existing credential', async () => {
      const credentialId: string = 'test-credential';
      fakeFilesystem[cacheFilePath] = JSON.stringify({
        version: '0.1.0',
        cacheEntries: {
          [credentialId]: {
            expires: 0,
            credential: 'test-value'
          }
        }
      });

      const credentialCache1: CredentialCache = await CredentialCache.initializeAsync({
        cacheFilePath: cacheFilePath,
        supportEditing: true
      });
      credentialCache1.deleteCacheEntry(credentialId);
      expect(credentialCache1.tryGetCacheEntry(credentialId)).toBeUndefined();
      await credentialCache1.saveIfModifiedAsync();
      credentialCache1.dispose();

      expect(fakeFilesystem[cacheFilePath]).toMatchSnapshot('credential cache file');

      const credentialCache2: CredentialCache = await CredentialCache.initializeAsync({
        cacheFilePath: cacheFilePath,
        supportEditing: false
      });
      expect(credentialCache2.tryGetCacheEntry(credentialId)).toBeUndefined();
      credentialCache2.dispose();
    });

    it('correctly sets credentialMetadata', async () => {
      const credentialId: string = 'test-credential';
      const credentialValue: string = 'test-value';
      const credentialMetadata: object = {
        a: 1,
        b: true
      };

      const credentialCache1: CredentialCache = await CredentialCache.initializeAsync({
        cacheFilePath: cacheFilePath,
        supportEditing: true
      });
      credentialCache1.setCacheEntry(credentialId, { credential: credentialValue, credentialMetadata });
      expect(credentialCache1.tryGetCacheEntry(credentialId)).toEqual({
        credential: credentialValue,
        credentialMetadata
      });
      await credentialCache1.saveIfModifiedAsync();
      credentialCache1.dispose();

      expect(fakeFilesystem[cacheFilePath]).toMatchSnapshot('credential cache file');

      const credentialCache2: CredentialCache = await CredentialCache.initializeAsync({
        cacheFilePath: cacheFilePath,
        supportEditing: false
      });
      expect(credentialCache2.tryGetCacheEntry(credentialId)).toEqual({
        credential: credentialValue,
        credentialMetadata
      });
      credentialCache2.dispose();
    });

    it('correctly updates credentialMetadata', async () => {
      const credentialId: string = 'test-credential';
      const credentialValue: string = 'test-value';
      const oldCredentialMetadata: object = {
        a: 1,
        b: true
      };
      const newCredentialMetadata: object = {
        c: ['a', 'b', 'c']
      };

      fakeFilesystem[cacheFilePath] = JSON.stringify({
        version: '0.1.0',
        cacheEntries: {
          [credentialId]: {
            expires: 0,
            credential: 'test-value',
            credentialMetadata: oldCredentialMetadata
          }
        }
      });

      const credentialCache1: CredentialCache = await CredentialCache.initializeAsync({
        cacheFilePath: cacheFilePath,
        supportEditing: true
      });
      credentialCache1.setCacheEntry(credentialId, {
        credential: credentialValue,
        credentialMetadata: newCredentialMetadata
      });
      expect(credentialCache1.tryGetCacheEntry(credentialId)).toEqual({
        credential: credentialValue,
        credentialMetadata: newCredentialMetadata
      });
      await credentialCache1.saveIfModifiedAsync();
      credentialCache1.dispose();

      expect(fakeFilesystem[cacheFilePath]).toMatchSnapshot('credential cache file');

      const credentialCache2: CredentialCache = await CredentialCache.initializeAsync({
        cacheFilePath: cacheFilePath,
        supportEditing: false
      });
      expect(credentialCache2.tryGetCacheEntry(credentialId)).toEqual({
        credential: credentialValue,
        credentialMetadata: newCredentialMetadata
      });
      credentialCache2.dispose();
    });
  });

  it('does not allow interaction if already disposed', async () => {
    const credentialCache: CredentialCache = await CredentialCache.initializeAsync({ supportEditing: true });
    credentialCache.dispose();

    expect(() => credentialCache.deleteCacheEntry('test')).toThrowErrorMatchingInlineSnapshot(
      `"This instance of CredentialCache has been disposed."`
    );
    await expect(() => credentialCache.saveIfModifiedAsync()).rejects.toThrowErrorMatchingInlineSnapshot(
      `"This instance of CredentialCache has been disposed."`
    );
    expect(() =>
      credentialCache.setCacheEntry('test', { credential: 'test' })
    ).toThrowErrorMatchingInlineSnapshot(`"This instance of CredentialCache has been disposed."`);
    expect(() => credentialCache.trimExpiredEntries()).toThrowErrorMatchingInlineSnapshot(
      `"This instance of CredentialCache has been disposed."`
    );
    expect(() => credentialCache.tryGetCacheEntry('test')).toThrowErrorMatchingInlineSnapshot(
      `"This instance of CredentialCache has been disposed."`
    );
  });

  it("does not allow modification if initialized with 'supportEditing': false", async () => {
    const credentialCache: CredentialCache = await CredentialCache.initializeAsync({ supportEditing: false });

    expect(() => credentialCache.deleteCacheEntry('test')).toThrowErrorMatchingInlineSnapshot(
      `"This instance of CredentialCache does not support editing."`
    );
    await expect(() => credentialCache.saveIfModifiedAsync()).rejects.toThrowErrorMatchingInlineSnapshot(
      `"This instance of CredentialCache does not support editing."`
    );
    expect(() =>
      credentialCache.setCacheEntry('test', { credential: 'test' })
    ).toThrowErrorMatchingInlineSnapshot(`"This instance of CredentialCache does not support editing."`);
    expect(() => credentialCache.trimExpiredEntries()).toThrowErrorMatchingInlineSnapshot(
      `"This instance of CredentialCache does not support editing."`
    );
  });
});
