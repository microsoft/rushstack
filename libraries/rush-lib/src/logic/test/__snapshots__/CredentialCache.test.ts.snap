// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CredentialCache cache paths [custom cache path with json suffix] correctly adds a new credential: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {
    \\"test-credential\\": {
      \\"expires\\": 0,
      \\"credential\\": \\"test-value\\"
    }
  }
}
"
`;

exports[`CredentialCache cache paths [custom cache path with json suffix] correctly deletes an existing credential: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {}
}
"
`;

exports[`CredentialCache cache paths [custom cache path with json suffix] correctly sets credentialMetadata: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {
    \\"test-credential\\": {
      \\"expires\\": 0,
      \\"credential\\": \\"test-value\\",
      \\"credentialMetadata\\": {
        \\"a\\": 1,
        \\"b\\": true
      }
    }
  }
}
"
`;

exports[`CredentialCache cache paths [custom cache path with json suffix] correctly trims expired credentials: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {}
}
"
`;

exports[`CredentialCache cache paths [custom cache path with json suffix] correctly updates an existing credential: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {
    \\"test-credential\\": {
      \\"expires\\": 0,
      \\"credential\\": \\"new-test-value\\"
    }
  }
}
"
`;

exports[`CredentialCache cache paths [custom cache path with json suffix] correctly updates credentialMetadata: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {
    \\"test-credential\\": {
      \\"expires\\": 0,
      \\"credential\\": \\"test-value\\",
      \\"credentialMetadata\\": {
        \\"c\\": [
          \\"a\\",
          \\"b\\",
          \\"c\\"
        ]
      }
    }
  }
}
"
`;

exports[`CredentialCache cache paths [custom cache path with json suffix] initializes a credential cache correctly when one exists on disk with a expired credential: expiration 1`] = `1970-01-01T00:00:00.100Z`;

exports[`CredentialCache cache paths [custom cache path with no suffix] correctly adds a new credential: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {
    \\"test-credential\\": {
      \\"expires\\": 0,
      \\"credential\\": \\"test-value\\"
    }
  }
}
"
`;

exports[`CredentialCache cache paths [custom cache path with no suffix] correctly deletes an existing credential: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {}
}
"
`;

exports[`CredentialCache cache paths [custom cache path with no suffix] correctly sets credentialMetadata: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {
    \\"test-credential\\": {
      \\"expires\\": 0,
      \\"credential\\": \\"test-value\\",
      \\"credentialMetadata\\": {
        \\"a\\": 1,
        \\"b\\": true
      }
    }
  }
}
"
`;

exports[`CredentialCache cache paths [custom cache path with no suffix] correctly trims expired credentials: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {}
}
"
`;

exports[`CredentialCache cache paths [custom cache path with no suffix] correctly updates an existing credential: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {
    \\"test-credential\\": {
      \\"expires\\": 0,
      \\"credential\\": \\"new-test-value\\"
    }
  }
}
"
`;

exports[`CredentialCache cache paths [custom cache path with no suffix] correctly updates credentialMetadata: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {
    \\"test-credential\\": {
      \\"expires\\": 0,
      \\"credential\\": \\"test-value\\",
      \\"credentialMetadata\\": {
        \\"c\\": [
          \\"a\\",
          \\"b\\",
          \\"c\\"
        ]
      }
    }
  }
}
"
`;

exports[`CredentialCache cache paths [custom cache path with no suffix] initializes a credential cache correctly when one exists on disk with a expired credential: expiration 1`] = `1970-01-01T00:00:00.100Z`;

exports[`CredentialCache cache paths [default cache path] correctly adds a new credential: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {
    \\"test-credential\\": {
      \\"expires\\": 0,
      \\"credential\\": \\"test-value\\"
    }
  }
}
"
`;

exports[`CredentialCache cache paths [default cache path] correctly deletes an existing credential: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {}
}
"
`;

exports[`CredentialCache cache paths [default cache path] correctly sets credentialMetadata: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {
    \\"test-credential\\": {
      \\"expires\\": 0,
      \\"credential\\": \\"test-value\\",
      \\"credentialMetadata\\": {
        \\"a\\": 1,
        \\"b\\": true
      }
    }
  }
}
"
`;

exports[`CredentialCache cache paths [default cache path] correctly trims expired credentials: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {}
}
"
`;

exports[`CredentialCache cache paths [default cache path] correctly updates an existing credential: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {
    \\"test-credential\\": {
      \\"expires\\": 0,
      \\"credential\\": \\"new-test-value\\"
    }
  }
}
"
`;

exports[`CredentialCache cache paths [default cache path] correctly updates credentialMetadata: credential cache file 1`] = `
"{
  \\"version\\": \\"0.1.0\\",
  \\"cacheEntries\\": {
    \\"test-credential\\": {
      \\"expires\\": 0,
      \\"credential\\": \\"test-value\\",
      \\"credentialMetadata\\": {
        \\"c\\": [
          \\"a\\",
          \\"b\\",
          \\"c\\"
        ]
      }
    }
  }
}
"
`;

exports[`CredentialCache cache paths [default cache path] initializes a credential cache correctly when one exists on disk with a expired credential: expiration 1`] = `1970-01-01T00:00:00.100Z`;
